pkcs11/work.go

```diff
@@ -26,68 +35,193 @@ type Work struct {
+				go w.work.signerData.signData(reqCtx, sResp.signer, w.work.pool, dataCh, errCh)
+			}
+		case resp := <-dataCh:
+			ht = time.Since(hStart).Nanoseconds() / time.Microsecond.Nanoseconds()
+			// Case 5: HSM has completed the signing operation & we need to send response back to client.
+			w.work.respChan <- resp
+			return
+		case err := <-errCh:
+			w.work.errChan <- err
+			return
 		}
 	}
 }
 
-// sendResponse sends the response on the respChan if the channel is not yet closed by the client.
-func (w *Work) sendResponse(resp signerWithSignAlgorithm) {
+// getData gets X509 CA certificate.
+func (s *signerX509) getData(ctx context.Context, signer signerWithSignAlgorithm, pool sPool, data chan []byte, errCh chan error) {
+
+}
+
+// signData signs X509 certificate by using the signer fetched from the pool.
+func (s *signerX509) signData(ctx context.Context, signer signerWithSignAlgorithm, pool sPool, data chan []byte, errCh chan error) {
+	defer pool.put(signer)
+	// Validate the cert request to ensure it matches the keyType and also the HSM supports the signature algo.
+	if val := isValidCertRequest(s.cert, signer.signAlgorithm()); !val {
+		log.Printf("signX509cert: cn=%q unsupported-sa=%q supported-sa=%d",
+			s.x509CACert.Subject.CommonName, s.cert.SignatureAlgorithm.String(), signer.signAlgorithm())
+		// Not a valid signature algorithm. Overwrite it with what the configured keyType supports.
+		s.cert.SignatureAlgorithm = signer.signAlgorithm()
+	}
+
+	s.cert.OCSPServer = s.ocspServer
+	s.cert.CRLDistributionPoints = s.crlDistribPoints
+
+	signedCert, err := x509.CreateCertificate(rand.Reader, s.cert, s.x509CACert, s.cert.PublicKey, signer)
+	if err != nil {
+		errCh <- err
+		return
+	}
+	select {
+	case <-ctx.Done():
+	case data <- pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: signedCert}):
+	}
+}
+
+// getData gets SSH certificate signing key by using the signer fetched from the pool.
+func (s *signerSSH) getData(ctx context.Context, signer signerWithSignAlgorithm, pool sPool, data chan []byte, errCh chan error) {
+	defer pool.put(signer)
+
+	sshSigner, err := ssh.NewSignerFromSigner(signer)
+	if err != nil {
+		errCh <- fmt.Errorf("failed to create sshSigner: %v", err)
+		return
+	}
+	data <- ssh.MarshalAuthorizedKey(sshSigner.PublicKey())
+}
+
+// signData signs SSH certificate by using the signer fetched from the pool.
+func (s *signerSSH) signData(ctx context.Context, signer signerWithSignAlgorithm, pool sPool, data chan []byte, errCh chan error) {
+	defer pool.put(signer)
+	if s.cert == nil {
+		errCh <- errors.New("signSSHCert: cannot sign empty cert")
+		return
+	}
+
+	sshSigner, err := newAlgorithmSignerFromSigner(signer, signer.publicKeyAlgorithm(), signer.signAlgorithm())
+	if err != nil {
+		errCh <- fmt.Errorf("failed to new ssh signer from signer, error :%v", err)
+		return
+	}
+	if err := s.cert.SignCert(rand.Reader, sshSigner); err != nil {
+		errCh <- err
+		return
+	}
+	select {
```

Comment:
Do we need select for `data <- bytes.TrimSpace(ssh.MarshalAuthorizedKey(s.cert))`?

Reply:
Yes, ctx.Done can guarantee `data` channel is available